programs/hyperplane/src/curve/stable.rs

```
@@ -216,14 +216,112 @@ fn compute_y(ann: u64, x: u128, d: u128) -> Result<u128> {
     u128::try_from(y).map_err(|_| error!(SwapError::CalculationFailure))
 }
 
+fn scale_up(source_amount: u128, factor: u64) -> Result<u128> {
+    let amount = if factor > 1 {
+        try_math!(source_amount.try_mul(factor as u128))?
+    } else {
+        source_amount
+    };
+    Ok(amount)
+}
+
+fn scale_down(source_amount: u128, factor: u64, round_up: bool) -> Result<u128> {
+    let amount = if factor > 1 {
+        let factor = u128::from(factor);
+        let amount = source_amount / factor;
+        // Was there any remainder?
+        if round_up && factor * amount < source_amount {
+            amount + 1
+        } else {
+            amount
+        }
+    } else {
+        source_amount
+    };
+    Ok(amount)
```

`factor == 0` is not handled and I think it should return an error. At the moment, you handle it as if `factor == 1`, and this is not right.

---

Added require_msg! on factor == 0